Readme.md
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
![CI](https://github.com/vitordelfino/express-handler-errors/workflows/CI/badge.svg?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/vitordelfino/express-handler-errors/badge.svg?branch=master)](https://coveralls.io/github/vitordelfino/express-handler-errors?branch=master)
![](https://img.shields.io/github/last-commit/vitordelfino/express-handler-errors/master)
![](https://img.shields.io/github/issues/vitordelfino/express-handler-errors/master)
![](https://img.shields.io/npm/dt/express-handler-errors)
![](https://img.shields.io/npm/v/express-handler-errors)
![](https://img.shields.io/github/package-json/keywords/vitordelfino/express-handler-errors)

# Express Handler Errors

Middleware to send custom errors in yours requests

## How to use

![configure](./docs/configure.png)

- Using with [Prisma v2](https://www.prisma.io/docs/)

![prisma](./docs/prisma.png)

You can pass a winston instance to log errors

![logger](./docs/logger.png)

In your service

![service](./docs/service.png)

## implement your own error functions and pass by parameter

- Error classes

![config class](./docs/config-class.png)

- Configure Hadler on express

![config functions](./docs/config-functions.png)


CHANGELOG.md
## [2.3.1](https://github.com/vitordelfino/express-handler-errors/compare/v2.3.0...v2.3.1) (2020-07-20)


### Bug Fixes

* remove console.log ([ed7bbc3](https://github.com/vitordelfino/express-handler-errors/commit/ed7bbc3b414d2a4a858a85cb99fccf13b3d49235))

# [2.3.0](https://github.com/vitordelfino/express-handler-errors/compare/v2.2.1...v2.3.0) (2020-07-20)


### Features

* **handler:** accept functions ([3e8154f](https://github.com/vitordelfino/express-handler-errors/commit/3e8154f6eb966c007080ce5baeb9c46152420b26))

## [2.2.1](https://github.com/vitordelfino/express-handler-errors/compare/v2.2.0...v2.2.1) (2020-07-15)


### Bug Fixes

* **custom-prisma-error:** extends error on class ([5df19b1](https://github.com/vitordelfino/express-handler-errors/commit/5df19b1d821bbd4a201890e710395844eb90344d))

# [2.2.0](https://github.com/vitordelfino/express-handler-errors/compare/v2.1.1...v2.2.0) (2020-07-14)


### Features

* **error-handler:** add prisma2 support ([c6692c5](https://github.com/vitordelfino/express-handler-errors/commit/c6692c51c2a9ba8d030be9220f2dd649bb3255c9))

## [2.1.1](https://github.com/vitordelfino/express-handler-errors/compare/v2.1.0...v2.1.1) (2020-07-09)


### Bug Fixes

* remove default export ([501d3d0](https://github.com/vitordelfino/express-handler-errors/commit/501d3d06cffb7824795d4b71001af1694442c17c))

# [2.1.0](https://github.com/vitordelfino/express-handler-errors/compare/v2.0.6...v2.1.0) (2020-07-09)


### Features

* try semantic-release ([4b903ac](https://github.com/vitordelfino/express-handler-errors/commit/4b903ac2a1653843d5cad2f8165562b0841c1c43))


test/setup.ts
import { plainToClass } from "class-transformer";
import { validateOrReject } from "class-validator";
import express, { Response, NextFunction } from "express";
import * as yup from "yup";

import { ErrorHandler } from "../src/lib/ErrorHandler";
import { CustomError } from "../src/models/CustomError";
import { CustomPrismaError } from "../src/models/CustomPrismaError";
import { Teste, ElseOtherCustomError, OtherCustomError } from "./ClassTest";
import logger from "./mock-logger";
import "express-async-errors";

const app = express();
const app2 = express();

app.use(express.json());
app2.use(express.json());

const yupError = async (req: any, _: Response, next: NextFunction) => {
  await yup
    .object()
    .shape({
      teste: yup.string().required()
    })
    .validate(req.body, { abortEarly: false });
  return next();
};

const classValidatorError = async (
  req: any,
  _: Response,
  next: NextFunction
) => {
  const transformed = plainToClass(Teste, req.body);
  await validateOrReject(transformed);
  return next();
};

app.post("/yup", yupError, (_: any, res: Response) => {
  res.json({ message: "OK" });
});

app.post("/class-validator", classValidatorError, (_, res: Response) => {
  res.json({ message: "OK" });
});

app.get("/custom-error", () => {
  throw new CustomError({
    code: "CustomError",
    message: "custom error",
    status: 400
  });
});

app.get("/custom-prisma-error", () => {
  throw new CustomPrismaError(["Some error", "Other error"]);
});

app.get("/error", () => {
  throw new Error("some error");
});

app2.get("/custom-other-error", () => {
  throw new OtherCustomError({
    code: "OTHER_ERROR",
    status: 400,
    message: "Throw another error"
  });
});

app2.get("/else-custom-other-error", () => {
  throw new ElseOtherCustomError({
    code: "ELSE_OTHER_ERROR",
    status: 500,
    message: "Throw another error"
  });
});

const otherError = (err: any, res: Response, next: NextFunction): any => {
  if (err instanceof OtherCustomError) {
    res.status(400).json(err.getErroResponse());
    return next(err);
  }
};

const elseOtherError = (err: any, res: Response, next: NextFunction): any => {
  if (err instanceof ElseOtherCustomError) {
    res.status(500).json(err.getErroResponse());
    return next(err);
  }
};

app.use((err: Error, _: any, res: Response, next: NextFunction) => {
  new ErrorHandler().handle(err, res, next);
});

app2.use((err: Error, _: any, res: Response, next: NextFunction) => {
  new ErrorHandler().handle(
    err,
    res,
    next,
    logger as any,
    otherError,
    elseOtherError
  );
});
export { app, app2 };


test/ClassTest.ts
import { IsString, IsDefined } from "class-validator";

export class Teste {
  @IsString({ message: "$property must be string" })
  @IsDefined({ message: "$property must be defined" })
  param1!: string;

  @IsString({ message: "$property must be string" })
  @IsDefined({ message: "$property must be defined" })
  param2!: string;
}

export class OtherCustomError extends Error {
  public readonly error: any;

  constructor(error: any) {
    super(error.message);
    this.error = error;
  }

  public getErroResponse(): { otherErrors: any[] } {
    return {
      otherErrors: [this.error]
    };
  }
}

export class ElseOtherCustomError extends Error {
  public readonly error: any;

  constructor(error: any) {
    super(error.message);
    this.error = error;
  }

  public getErroResponse(): { elseOtherErrors: any[] } {
    return {
      elseOtherErrors: [this.error]
    };
  }
}


test/index.test.ts
import request from "supertest";

import app from "./setup";

describe("Error Handler Tests", () => {
  test("should return error when validate with yup", async () => {
    await request(app)
      .post("/yup")
      .expect(400, {
        errors: [
          { code: "ValidationError", message: "teste is a required field" }
        ]
      });
  });

  test("should return error when validate with class-validator", async () => {
    await request(app)
      .post("/class-validator")
      .send({ param1: "param" })
      .expect(400, {
        errors: [
          {
            code: "ValidationError",
            messages: ["param2 must be defined", "param2 must be string"]
          }
        ]
      });
  });

  test("should returm CustomError", async () => {
    await request(app)
      .get("/custom-error")
      .expect(400, {
        errors: [{ code: "CustomError", message: "custom error", status: 400 }]
      });
  });
  test("should returm CustomPrismaError", async () => {
    await request(app)
      .get("/custom-prisma-error")
      .expect(422, {
        code: "PrismaValidationError",
        errors: ["Some error", "Other error"],
        status: 422
      });
  });
  test("should returm Error", async () => {
    await request(app)
      .get("/error")
      .send({ param1: "param" })
      .expect(500, {
        errors: [{ code: "E0001", message: "some error" }]
      });
  });
});


.eslintrc.js
module.exports = {
  env: {
    es6: true,
    node: true,
    commonjs: true,
    jest: true,
  },
  extends: [
    'plugin:@typescript-eslint/recommended', 'airbnb-base', 'prettier', 'prettier/@typescript-eslint'
  ],
  globals: {
    Atomics: 'readonly',
    SharedArrayBuffer: 'readonly',
  },
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: 'module',
  },
  plugins: [
    '@typescript-eslint',
    'prettier',
    'eslint-plugin-import-helpers'
  ],
  rules: {
    'class-methods-use-this': 0,
    'no-unused-vars': ['error', { argsIgnorePattern: '[next|^_]' }],
    'no-console': 0,
    camelcase: 'off',
    "import/no-unresolved": "off",
    'import/prefer-default-export': 'off',
    'no-param-reassign': 'off',
    '@typescript-eslint/camelcase': ['off', { ignoreDestructuring: true }],
    'import/extensions': [
      'error',
      'ignorePackages',
      {
        js: 'never',
        ts: 'never',
      },
    ],
  },
  settings: {
    'import/resolver': {
      typescript: {
        directory: './tsconfig.json',
      },
      node: {
        paths: ['src'],
        extensions: ['.ts'],
      },
    },
  }
};


src/lib/ErrorHandler.ts
import { ValidationError } from "class-validator";
import { Response, NextFunction } from "express";
import { Logger } from "winston";
import { ValidationError as YupValidationError } from "yup";

import { CustomError } from "../models/CustomError";
import { CustomPrismaError } from "../models/CustomPrismaError";

export class ErrorHandler {
  public handle(
    err: Error,
    res: Response,
    next: NextFunction,
    logger?: Logger,
    ...callbacks: Function[]
  ): void {
    if (logger)
      logger.error(`ErrorHandle::handle::${err.name}::${err.message}`);
    if (callbacks && callbacks.length > 0) {
      callbacks.forEach(callback => callback(err, res, next));
    }
    if (err instanceof CustomPrismaError) {
      res.status(err.status).json(err.getErrorResponse());
    } else if (err instanceof CustomError) {
      res.status(err.error.status).json(err.getErrorResponse());
    } else if (err instanceof YupValidationError) {
      const errors = err.errors.map((error: string) => ({
        code: err.name,
        message: error
      }));
      res.status(400).json({ errors });
    } else if (Array.isArray(err) && err[0] instanceof ValidationError) {
      const errors = err.map((e: ValidationError) => {
        const messages = Object.keys(e.constraints!).map(
          key => e.constraints![key]
        );

        return { code: "ValidationError", messages };
      });
      res.status(400).json({ errors });
    } else {
      res.status(500).json({
        errors: [
          {
            code: "E0001",
            message: err.message
          }
        ]
      });
    }
    next(err);
  }
}
